File: emscripten/npm/src/VerovioToolkit.js
import VerovioEmscriptenAdapter from './VerovioEmscriptenAdapter';

const VerovioToolkit = class {

    constructor() {
        this.ptr = VerovioEmscriptenAdapter.vrvToolkit.constructor();
        console.debug("Creating toolkit instance");
        VerovioEmscriptenAdapter.instances.push(this.ptr);
    }

    destroy() {
        VerovioEmscriptenAdapter.instances.splice(VerovioEmscriptenAdapter.instances.indexOf(this.ptr), 1);
        console.debug("Deleting toolkit instance");
        VerovioEmscriptenAdapter.vrvToolkit.destructor(this.ptr);
    }

    edit(editorAction) {
        return VerovioEmscriptenAdapter.vrvToolkit.edit(this.ptr, JSON.stringify(editorAction));
    }

    editInfo() {
        return JSON.parse(VerovioEmscriptenAdapter.vrvToolkit.editInfo(this.ptr));
    }

    getAvailableOptions() {
        return JSON.parse(VerovioEmscriptenAdapter.vrvToolkit.getAvailableOptions(this.ptr));
    }

    getElementAttr(xmlId) {
        return JSON.parse(VerovioEmscriptenAdapter.vrvToolkit.getElementAttr(this.ptr, xmlId));
    }

    getElementsAtTime(millisec) {
        return JSON.parse(VerovioEmscriptenAdapter.vrvToolkit.getElementsAtTime(this.ptr, millisec));
    }

    getExpansionIdsForElement(xmlId) {
        return JSON.parse(VerovioEmscriptenAdapter.vrvToolkit.getExpansionIdsForElement(this.ptr, xmlId));
    }

    getHumdrum() {
        return VerovioEmscriptenAdapter.vrvToolkit.getHumdrum(this.ptr);
    }

    getLog() {
        return VerovioEmscriptenAdapter.vrvToolkit.getLog(this.ptr);
    }

    getMEI(pageNo, scoreBased) {
        return VerovioEmscriptenAdapter.vrvToolkit.getMEI(this.ptr, pageNo, scoreBased);
    }

    getMIDIValuesForElement(xmlId) {
        return JSON.parse(VerovioEmscriptenAdapter.vrvToolkit.getMIDIValuesForElement(this.ptr, xmlId));
    }

    getNotatedIdForElement(xmlId) {
        return VerovioEmscriptenAdapter.vrvToolkit.getNotatedIdForElement(this.ptr, xmlId);
    }

    getOptions(defaultValues) {
        return JSON.parse(VerovioEmscriptenAdapter.vrvToolkit.getOptions(this.ptr, defaultValues));
    }

    getPageCount() {
        return VerovioEmscriptenAdapter.vrvToolkit.getPageCount(this.ptr);
    }

    getPageWithElement(xmlId) {
        return VerovioEmscriptenAdapter.vrvToolkit.getPageWithElement(this.ptr, xmlId);
    }

    getTimeForElement(xmlId) {
        return VerovioEmscriptenAdapter.vrvToolkit.getTimeForElement(this.ptr, xmlId);
    }

    getVersion() {
        return VerovioEmscriptenAdapter.vrvToolkit.getVersion(this.ptr);
    }

    loadData(data) {
        return VerovioEmscriptenAdapter.vrvToolkit.loadData(this.ptr, data);
    }

    redoLayout() {
        VerovioEmscriptenAdapter.vrvToolkit.redoLayout(this.ptr);
    }

    redoPagePitchPosLayout() {
        VerovioEmscriptenAdapter.vrvToolkit.redoPagePitchPosLayout(this.ptr);
    }

    renderData(data, options) {
        return VerovioEmscriptenAdapter.vrvToolkit.renderData(this.ptr, data, JSON.stringify(options));
    }

    renderPage(pageNo, options) {
        console.warn("Method renderPage is deprecated; use renderToSVG instead");
        return VerovioEmscriptenAdapter.vrvToolkit.renderToSVG(this.ptr, pageNo, JSON.stringify(options));
    }

    renderToMIDI(options) {
        return VerovioEmscriptenAdapter.vrvToolkit.renderToMIDI(this.ptr, JSON.stringify(options));
    }

    renderToMidi(options) {
        console.warn("Method renderToMidi is deprecated; use renderToMIDI instead");
        return VerovioEmscriptenAdapter.vrvToolkit.renderToMIDI(this.ptr, JSON.stringify(options));
    }

    renderToSVG(pageNo, options) {
        return VerovioEmscriptenAdapter.vrvToolkit.renderToSVG(this.ptr, pageNo, JSON.stringify(options));
    }

    renderToTimemap() {
        return JSON.parse(VerovioEmscriptenAdapter.vrvToolkit.renderToTimemap(this.ptr));
    }

    setOptions(options) {
        VerovioEmscriptenAdapter.vrvToolkit.setOptions(this.ptr, JSON.stringify(options));
    }

};

// If the window object is defined (if we are not within a WebWorker)...
if ((typeof window !== "undefined") && (window.addEventListener)) {
    // Add a listener that will delete the object (if necessary) when the page is closed
    window.addEventListener("unload", function () {
        for (var i = 0; i < VerovioEmscriptenAdapter.instances.length; i++) {
            VerovioEmscriptenAdapter.vrvToolkit.destructor(VerovioEmscriptenAdapter.instances[i]);
        }
    });
}

export default VerovioToolkit;

File: bindings/swift-toolkit/VerovioToolkit.swift

import Foundation
import VerovioCore

public class VerovioToolkit {
    private var toolkitPtr: UnsafeMutableRawPointer?

    public init() {
        self.toolkitPtr = vrvToolkit_constructorNoResource()
    }
    
    public init(_ resourcePath: String) {
        resourcePath.withCString{ cStr in
            self.toolkitPtr = vrvToolkit_constructorResourcePath(cStr)
        }
    }

    deinit {
        if let tk = toolkitPtr {
            vrvToolkit_destructor(tk)
        }
    }
    
    public func edit(_ action: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return action.withCString { cStr in
            vrvToolkit_edit(tk, cStr)
        }
    }
    
    public func editInfo() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_editInfo(tk) else { return "" }
        return String(cString: cStr)
    }
    
    public func getAvailableOptions() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_getAvailableOptions(tk) else { return "" }
        return String(cString: cStr)
    }
    
    public func getDefaultOptions() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_getDefaultOptions(tk) else { return "" }
        return String(cString: cStr)
    }
    
    public func getDescriptiveFeatures(_ options: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return options.withCString { cStr in
            if let resultCStr = vrvToolkit_getDescriptiveFeatures(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }
    
    public func getElementAttr(_ element: String, _ attribute: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return element.withCString { elementCStr in
            if let resultCStr = vrvToolkit_getElementAttr(tk, elementCStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }
    
    public func getElementsAtTime(_ millisec: Int) -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_getElementsAtTime(tk, Int32(millisec)) else { return "" }
        return String(cString: cStr)
    }
    
    public func getExpansionIdsForElement(_ xmlId: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return xmlId.withCString { cStr in
            if let resultCStr = vrvToolkit_getExpansionIdsForElement(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }

    public func getHumdrum() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_getHumdrum(tk) else { return "" }
        return String(cString: cStr)
    }

    public func getHumdrumFile(_ filename: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return filename.withCString { cStr in
            vrvToolkit_getHumdrumFile(tk, cStr)
        }
    }

    public func getID() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_getID(tk) else { return "" }
        return String(cString: cStr)
    }

    public func convertHumdrumToHumdrum(_ humdrumData: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return humdrumData.withCString { cStr in
            if let resultCStr = vrvToolkit_convertHumdrumToHumdrum(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }

    public func convertHumdrumToMIDI(_ humdrumData: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return humdrumData.withCString { cStr in
            if let resultCStr = vrvToolkit_convertHumdrumToMIDI(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }

    public func convertMEIToHumdrum(_ meiData: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return meiData.withCString { cStr in
            if let resultCStr = vrvToolkit_convertMEIToHumdrum(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }

    public func getLog() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_getLog(tk) else { return "" }
        return String(cString: cStr)
    }

    public func getMEI(_ options: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return options.withCString { cStr in
            if let resultCStr = vrvToolkit_getMEI(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }

    public func getMIDIValuesForElement(_ xmlId: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return xmlId.withCString { cStr in
            if let resultCStr = vrvToolkit_getMIDIValuesForElement(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }

    public func getNotatedIdForElement(_ xmlId: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return xmlId.withCString { cStr in
            if let resultCStr = vrvToolkit_getNotatedIdForElement(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }

    public func getOptions() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_getOptions(tk) else { return "" }
        return String(cString: cStr)
    }

    public func getOptionUsageString() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_getOptionUsageString(tk) else { return "" }
        return String(cString: cStr)
    }
    
    public func getPageCount() -> Int {
        guard let tk = toolkitPtr else { return 0 }
        return Int(vrvToolkit_getPageCount(tk))
    }

    public func getPageWithElement(_ xmlId: String) -> Int {
        guard let tk = toolkitPtr else { return 0 }
        return xmlId.withCString { cStr in
            Int(vrvToolkit_getPageWithElement(tk, cStr))
        }
    }

    public func getResourcePath() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_getResourcePath(tk) else { return "" }
        return String(cString: cStr)
    }

    public func getScale() -> Int {
        guard let tk = toolkitPtr else { return 0 }
        return Int(vrvToolkit_getScale(tk))
    }

    public func getTimeForElement(_ xmlId: String) -> Double {
        guard let tk = toolkitPtr else { return 0 }
        return xmlId.withCString { cStr in
            vrvToolkit_getTimeForElement(tk, cStr)
        }
    }

    public func getTimesForElement(_ xmlId: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return xmlId.withCString { cStr in
            if let resultCStr = vrvToolkit_getTimesForElement(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }
    
    public func getVersion() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_getVersion(tk) else { return "" }
        return String(cString: cStr)
    }
    
    public func loadData(_ data: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return data.withCString { cStr in
            vrvToolkit_loadData(tk, cStr)
        }
    }

    public func loadFile(_ filename: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return filename.withCString { cStr in
            vrvToolkit_loadFile(tk, cStr)
        }
    }

    public func loadZipDataBase64(_ base64: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return base64.withCString { cStr in
            vrvToolkit_loadZipDataBase64(tk, cStr)
        }
    }
    
    public func redoLayout(_ jsonOptions: String) {
        guard let tk = toolkitPtr else { return }
        jsonOptions.withCString { cStr in
            vrvToolkit_redoLayout(tk, cStr)
        }
    }

    public func redoPagePitchPosLayout() {
        guard let tk = toolkitPtr else { return }
        vrvToolkit_redoPagePitchPosLayout(tk)
    }

    public func renderToExpansionMap() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_renderToExpansionMap(tk) else { return "" }
        return String(cString: cStr)
    }

    public func renderToExpansionMapFile(_ filename: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return filename.withCString { cStr in
            vrvToolkit_renderToExpansionMapFile(tk, cStr)
        }
    }

    public func renderToMIDI() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_renderToMIDI(tk) else { return "" }
        return String(cString: cStr)
    }

    public func renderToMIDIFile(_ filename: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return filename.withCString { cStr in
            vrvToolkit_renderToMIDIFile(tk, cStr)
        }
    }

    public func renderToPAE() -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_renderToPAE(tk) else { return "" }
        return String(cString: cStr)
    }

    public func renderToPAEFile(_ filename: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return filename.withCString { cStr in
            vrvToolkit_renderToPAEFile(tk, cStr)
        }
    }
    
    
    public func renderToSVG(_ pageNo: Int, _ xmlDeclaration: Bool) -> String {
        guard let tk = toolkitPtr else { return "" }
        guard let cStr = vrvToolkit_renderToSVG(tk, Int32(pageNo), xmlDeclaration) else { return "" }
        return String(cString: cStr)
    }
    
    public func renderToSVGFile(_ filename: String, _ pageNo: Int) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return filename.withCString { cStr in
            vrvToolkit_renderToSVGFile(tk, cStr, Int32(pageNo))
        }
    }

    public func renderToTimemap(_ options: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return options.withCString { cStr in
            if let resultCStr = vrvToolkit_renderToTimemap(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }

    public func renderToTimemapFile(_ filename: String, _ options: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return filename.withCString { fStr in
            options.withCString { oStr in
                vrvToolkit_renderToTimemapFile(tk, fStr, oStr)
            }
        }
    }

    public func resetOptions() {
        guard let tk = toolkitPtr else { return }
        vrvToolkit_resetOptions(tk)
    }

    public func resetXmlIdSeed(_ seed: Int) {
        guard let tk = toolkitPtr else { return }
        vrvToolkit_resetXmlIdSeed(tk, Int32(seed))
    }

    public func saveFile(_ filename: String, _ options: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return filename.withCString { fStr in
            options.withCString { oStr in
                vrvToolkit_saveFile(tk, fStr, oStr)
            }
        }
    }

    public func select(_ selection: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return selection.withCString { cStr in
            vrvToolkit_select(tk, cStr)
        }
    }

    public func setInputFrom(_ input: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return input.withCString { cStr in
            vrvToolkit_setInputFrom(tk, cStr)
        }
    }
    
    
    public func setOptions(_ jsonOptions: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return jsonOptions.withCString { cStr in
            vrvToolkit_setOptions(tk, cStr)
        }
    }

    public func setOutputTo(_ output: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return output.withCString { cStr in
            vrvToolkit_setOutputTo(tk, cStr)
        }
    }
    
    public func setResourcePath(_ path: String) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return path.withCString { cStr in
            vrvToolkit_setResourcePath(tk, cStr)
        }
    }

    public func setScale(_ scale: Int) -> Bool {
        guard let tk = toolkitPtr else { return false }
        return vrvToolkit_setScale(tk, Int32(scale))
    }

    public func validatePAE(_ data: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return data.withCString { cStr in
            if let resultCStr = vrvToolkit_validatePAE(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }

    public func validatePAEFile(_ filename: String) -> String {
        guard let tk = toolkitPtr else { return "" }
        return filename.withCString { cStr in
            if let resultCStr = vrvToolkit_validatePAEFile(tk, cStr) {
                return String(cString: resultCStr)
            } else {
                return ""
            }
        }
    }
}
